### PR TITLE
The hex_to_bin() is replaced by sc_hex_to_bin()

### PR DESCRIPTION
**Pull request by popovec**
It is not advisable to have two functions that perform the same operation.
The hex_to_bin() is used only in pkcs11-tool, but some cases generates wrong
output. This function is removed in favor of the sc_hex_to_bin() function.

Fixes: #2465

### CODE REVIEW

**Comment by frankmorgner** [comment]
Thank you!

Would it be possible to write a test case that verifies fixing the parsing issue?

File: src/tests/unittests/hextobin.c [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+#include <string.h>
+#include "libopensc/opensc.h"
+
+#define LEN 15
+
+int main()
+{
+	char *t;
+	char test[][LEN] = {
+/* result length (0x60 error string, 0x70 end of array), string, result string */
```
Comment (Jakuje) [inline src/tests/unittests/hextobin.c:31]: I would suggest to use some struct instead of two-dimensional array with comment to describe its functionality. Something like this would make both the structure and the code more readable: 
```
struct {
	int result_len;
	const char *input;
	const char *output;
} test[] = {
	1, "0", "\x00",
	1, ":0", "\x00",
	...
};
```
(or even a bit more verbose with separating results and expected output length with proper types)

File: src/tests/unittests/hextobin.c [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+		{1}, "10", "\x10",
+		{1}, " 90", "\x90",
+		{1}, ":a0", "\xa0",
+		{1}, " :B0 :", "\xb0",
+
+		{1}, " 11:", "\x11",
+		{1}, " :11:", "\x11",
+		{1}, " :1 1:", "\x11",
+		{1}, "1:1", "\x11",
```
Comment (Jakuje) [inline src/tests/unittests/hextobin.c:60]: I am wondering if this is the expected output. I would expect `\x01\x01`, but it might be a matter of convention/taste/mood.

File: src/tests/unittests/hextobin.c [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+		{0}, ":", "",
+		{0}, " ", "",
+		{0}, "", "",
+
+		/* incorrect strings */
+		{0x60}, "111", "",
+		{0x60}, "b:211 :2", "",
+		{0x60}, "G", "",
+		{0x60}, " z", "",
+		{0x60}, "1:11", "",
```
Comment (Jakuje) [inline src/tests/unittests/hextobin.c:85]: I am wondering whats on `1:11` more wrong than on `1:234:56`, which is supposed to be a valid input.

**Comment by Jakuje** [comment]
Please, see @dengert [comment in #2465](https://github.com/OpenSC/OpenSC/issues/2465#issuecomment-1003041806). I think this function needs a bit more love. There are obviously some odds in the parsing and some expectations, which might not be completely clear for the users and which cause issues such as #2465. The least we should have a comments in the unit tests saying why the particular questionable corner cases are "correct" or "wrong" so next time somebody will question some decisions or change the behavior, we will have some reference.

File: src/libopensc/sc.c [hunk 4]
```diff
@@ -72,6 +72,13 @@ int sc_hex_to_bin(const char *in, u8 *out, size_t *outlen)
 	int r = SC_SUCCESS;
 	size_t left = *outlen;
 	u8 byte = 0;
+	/* skip initial separators */
```
Comment (frankmorgner) [inline src/libopensc/sc.c:75]: Initial seperators should be covered in the loop. In the first run byte_needs_nibble is 0 so that the loop will be continued in line 97, won't it?

**Review by frankmorgner — APPROVED** [review]
Great work, thanks for diving in so deep :wink:

**Review by Jakuje — APPROVED** [review]
